src/olympia/accounts/utils.py

```diff
@@ -32,18 +32,37 @@ def fxa_login_url(
     }
     if action is not None:
         query['action'] = action
-    if force_two_factor is True:
+    if enforce_two_factor_authentication is True:
         # Specifying AAL2 will require the token to have an authentication
         # assurance level >= 2 which corresponds to requiring 2FA.
         query['acr_values'] = 'AAL2'
-        # Requesting 'prompt=none' during authorization, together with passing
-        # a valid id token in 'id_token_hint', allows the user to not have to
-        # re-authenticate with FxA if they still have a valid session (which
-        # they should here: they went through FxA, back to AMO, and now we're
-        # redirecting them to FxA because we want them to have 2FA enabled).
-        if id_token:
+        # Requesting 'prompt=none' during authorization allows the user to not
+        # have to re-authenticate with FxA if they still have a valid session,
+        # in case they just went through FxA, then back to AMO only to find out
+        # we are requiring 2FA. To let FxA know who the user should be, we pass
+        # id_token_hint, which we have in this specific case.
+        # If we didn't just go through authentication, that means the user was
+        # alraedy logged in but we enforced 2FA at a later stage. We don't have
+        # id_token_hint, but we can pass login_hint instead.
+        # FIXME: this doesn't work, getting redirected to our callback with an
+        # unauthorized_client error from FxA. According to the docs usage of
+        # id_token_hint is always allowed but login_hint isn't, so we need to
+        # have our oauth client configs be explicitly allowed.
+        # https://mozilla.github.io/ecosystem-platform/reference/oauth-details
+        # #promptnone-support
+        # FIXME: this highlights the fact that we need to handle error cases
+        # and regenerate a new state to start from scratch - right now we're
+        # ignoring them, so the user can't recover. I'm not sure what's
+        # supposed to happen if the user goes to FxA with prompt=none and
+        # login_hint=email if they no longer have a valid FxA session (ideally
+        # FxA would handle that gracefully), but if an error is raised we'll
+        # need to handle it.
```

Comment:
We should follow https://mozilla.github.io/ecosystem-platform/reference/oauth-details#recovering-from-errors and log the error code, remove `prompt=none`, generate a new state and try again to handle errors.

Reply:
Filed https://github.com/mozilla/addons-server/issues/21002 to do this